update search/idastar (#94)
OI-wiki/i18n: docs/search/idastar.md

end;
```

### Advantage

1. The space overhead is small, and deep down is actually a depth-first search with limited depth. Also, it is well known that DFS consumes smaller space.
2. Beneficial for deep pruning.

### Disadvantage

Repetitive search: In the process of backtracking, each time the depth increases, the search must be done again from the beginning.

> > In fact, the difference between the previous search and the next one is negligible.

Reviewer: adjacent searches?

## Sample problem

??? note "Egypt fraction"
     **Problem description**

>     In ancient Egypt, people use the sum of unit fractions (that is, $\frac{1}{a}$ , $a$ is a natural number) to represent all rational numbers. For example, $\frac{2}{3}=\frac{1}{2}+\frac{1}{6}$ but $\frac{2}{3}=\frac{1}{ 3}+\frac{1}{3}$ is not allowed because same number cannot be added.

Reviewer: where $a\in\mathbb{N}^*$

>     In ancient Egypt, people use the sum of unit fractions (that is, $\frac{1}{a}$ , $a$ is a natural number) to represent all rational numbers. For example, $\frac{2}{3}=\frac{1}{2}+\frac{1}{6}$ but $\frac{2}{3}=\frac{1}{ 3}+\frac{1}{3}$ is not allowed because same number cannot be added.

Reviewer: ```suggestion
    In ancient Egypt, people use the sum of unit fractions (that is, $\frac{1}{a}$ , $a$ is a natural number) to represent all rational numbers. For example, $\frac{2}{3}=\frac{1}{2}+\frac{1}{6}$ but $\frac{2}{3}=\frac{1}{3}+\frac{1}{3}$ is not allowed because same numbers cannot be added.
```

>     For a fraction $\frac{a}{b}$ , there are many ways to represent it. Among them, the one with less addends is better than the one with more addends. If the number of additions is the same, the larger the minimum fraction, the better. For example, $\frac{19}{45}=\frac{1}{5}+\frac{1}{6}+\frac{1}{18}$ is the optimal solution.

Reviewer: Among them, the one with less operands is preferred, and when the number of operands equals, the larger ..., the better.

>     Enter integers $a,b$ ( $0<a<b<500$ ), and try to calculate the best expression.

Reviewer: Input

    sample input:

    ```text
    495 499
    ```

>     Sample output:

Reviewer: 大小写统一

    ```text
    Case 1: 495/499=1/2+1/5+1/6+1/8+1/3992+1/14970
    ```

 **Analysis** 

Theoretically this problem can be solved by the backtracking method, but the **solution tree** would be a "horror" - not only the depth has no obvious upper bound, but the choice of addends is theoretically unlimited. In other words, if BFS is used, even one layer cannot be expanded, because each layer is **infinite**.

> The solution is to use iterative deepening search: enumerate the upper limit of depth $maxd$ from small to large, and only consider nodes whose depth does not exceed $maxd$ for each execution. In this way, as long as the depth of the solution is optimized, it can be enumerated in a limited time.

Reviewer: \mathit{maxd}

Reviewer: 优先 | optimized 不太对

Author: prioritized? 🤔

> The maximum depth $maxd$ can also be used for **pruning**. The expansion is carried out in the order of increasing denominator. If it is expanded to level $i$ , the sum of the first $i$th fractions is $\frac{c}{d}$ , and the $i$th fraction is $\frac{1} {e}$ , then at least $\frac{\frac{a}{b}-\frac{c}{d}}{\frac{1}{e}}$ fractions are needed to achieve the total $\frac{a}{b}$ . For example, if $\frac{19}{45}=\frac{1}{5}+\frac{1}{100}+\cdots$ is currently searched, the largest element in the following fractions is up to $\frac{1 }{101}$ , and at least $\frac{\frac{19}{45}-\frac{1}{5}}{\frac{1}{101}}=23$ is required to reach the sum of $\frac {19}{45}$ . So the first $22$ iterations will not consider this subtree at all. The key point here is that it can be estimated at least how many more steps are needed to solve the problem.

Reviewer: ```suggestion
The maximum depth $maxd$ can also be used for **pruning**. The expansion is performed in the order of increasing denominators. If it is expanded to level $i$ , the sum of the first $i$-th fractions are $\frac{c}{d}$ , and the $i$-th fraction is $\frac{1} {e}$ , then at least $\frac{\frac{a}{b}-\frac{c}{d}}{\frac{1}{e}}$ fractions are needed to achieve the total of $\frac{a}{b}$ . For example, if $\frac{19}{45}=\frac{1}{5}+\frac{1}{100}+\cdots$ is currently searched, the largest element in the following fractions is up to $\frac{1}{101}$ , and at least $\frac{\frac{19}{45}-\frac{1}{5}}{\frac{1}{101}}=23$ is required to reach the sum of $\frac {19}{45}$ . So the first $22$ iterations will not consider this subtree at all. The key point here is that the least number of steps to solve the problem can be estimated.
```

LL gcd(LL a, LL b) { return b == 0 ? a : gcd(b, a % b); }

> // return satisfy the minimum c in 1/c <= a/b

Reviewer: ```suggestion
// return the minimum c which satisfies 1/c <= a/b
```

  return false;
}

// the current depth is d; the denominator cannot be less than `from`, and the sum of fractions is exactly aa/bb
bool dfs(int d, int from, LL aa, LL bb) {
  if (d == maxd) {
    if (bb % aa) return false;  // aa/bb must be the Egypt fraction
    v[d] = bb / aa;
    if (better(d)) memcpy(ans, v, sizeof(LL) * (d + 1));
    return true;
  }
  bool ok = false;
  from = max(from, get_first(aa, bb));  // beginning of enumeration
  for (int i = from;; i++) {
    // pruning: if the remaining maxd+1-d fractions are all 1/i, and the sum still does not exceed aa/bb, there is no solution
    if (bb * (maxd + 1 - d) <= i * aa) break;
    v[d] = i;
    // calculate aa/bb-1/i and set the result as a2/b2
    LL b2 = bb * i;
    LL a2 = aa * i - bb;

>     LL g = gcd(a2, b2);  // in order to reduce

Reviewer: reduce the fraction?

}
```

## Practice problems

>  [The rotation game UVa1343](https://www.luogu.com.cn/problem/UVA1343)(original link in Chinese)

Reviewer: https://onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&category=446&page=show_problem&problem=4089

## Practice problems

>  [The rotation game UVA1343](https://onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&category=446&page=show_problem&problem=4089)(original link in Chinese)

Reviewer: ```suggestion
 [The rotation game UVA1343](https://onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&category=446&page=show_problem&problem=4089)
```
